### DISCUSSION TITLE
gpio.pin.digital fail to lock GPIO04

### DISCUSSION THREAD

**Discussion by guitarddns** [CrazyIvan359/mqttany#148] [Troubleshooting]
Hi and thx for a great tool - a question about gpio part
I have a raspberry pi that I want to test the gpio part with mqttany - but receive an error when starting the service

```text
root@Infra-read:/home/john# systemctl status mqttany
â mqttany.service - MQTTany service
   Loaded: loaded (/etc/systemd/system/mqttany.service; enabled; vendor preset: enabled)
   Active: active (running) since Tue 2022-12-13 10:31:47 GMT; 9min ago
 Main PID: 427 (python3)
    Tasks: 11 (limit: 1935)
   CGroup: /system.slice/mqttany.service
           ââ427 /usr/bin/python3 mqttany.py
           ââ439 /usr/bin/python3 mqttany.py
           ââ440 /usr/bin/python3 mqttany.py
```

```log
Dec 13 10:31:49 Infra-read python3[427]: 2022-12-13 10:31:49,658 [WARN ] [core                    ] #                                     #
Dec 13 10:31:49 Infra-read python3[427]: 2022-12-13 10:31:49,659 [WARN ] [core                    ] #######################################
Dec 13 10:31:49 Infra-read python3[427]: 2022-12-13 10:31:49,661 [WARN ] [core                    ]
Dec 13 10:31:50 Infra-read python3[427]: 2022-12-13 10:31:50,007 [INFO ] [core.gpio               ] Detected board: RASPBERRY_PI_2B
Dec 13 10:31:50 Infra-read python3[427]: 2022-12-13 10:31:50,738 [WARN ] [core.gpio.board         ] 'gpio.pin.digital' failed to lock GPIO04 (GPIO04), already locked by 'system.1wire
Dec 13 10:31:50 Infra-read python3[427]: '
Dec 13 10:31:50 Infra-read python3[427]: 2022-12-13 10:31:50,756 [INFO ] [core                    ] Module 'mqtt' started successfully
Dec 13 10:31:50 Infra-read python3[427]: 2022-12-13 10:31:50,769 [INFO ] [core                    ] Module 'gpio' started successfully
Dec 13 10:31:50 Infra-read python3[427]: 2022-12-13 10:31:50,784 [WARN ] [gpio.digital            ] Unable to setup 'pin-id' as INPUT, no available pin class
Dec 13 10:31:50 Infra-read python3[427]: 2022-12-13 10:31:50,815 [INFO ] [mqtt                    ] Connected to broker '192.168.1.109:1883'
```

```text
root@Infra-read:/home/john# tail -10 /boot/config.txt
dtoverlay=vc4-fkms-v3d
max_framebuffers=2

[all]
#dtoverlay=vc4-fkms-v3d
start_x=1
gpu_mem=128
disable_camera_led=1
enable_uart=1
#dtoverlay=w1-gpio
```

```text
root@Infra-read:/home/john# ls /sys/bus/w1
ls: cannot access '/sys/bus/w1': No such file or directory
```

___________________________________________________________________

what could it be that is locking "system.1wire" 
thanks in advance

**Comment by CrazyIvan359** [comment]
I'm not entirely sure what you're trying to do here, can you post your config file? It looks like maybe you are trying to use the GPIO module to control the same pin that you are using for OneWire.

**Comment by guitarddns** [comment]
Hi Michael

Sorry for the confusion - I have 2 raspberry pi in place 
1 is running OneWire ( that one is solved now and work just fine  ) 
1 is running GPIO - I trying to address the pin via mqttany and have tried pigpio but no luck - and since the Onewire work out I try the GPIO also 

here is the mqttany.yml

***@***.***:~ $ more /etc/mqttany/mqttany.yml 
######## Configuration for MQTTany ########

# All optional values are commented out and default values are shown



### Config file version
version: '1.0.1'


######## MQTT Module ########
mqtt:

  ### Broker Host
  # can be either IPv4 address or hostname
  host: ‘192.168.x,x'

  ### Broker Port
  # port to connect to on the host
  port: 1883

  ### Client ID
  # Substitutions:
  # - {hostname} = the computer's hostname
  #client id: '{hostname}'

  ### Login
  username: mqttname
  password: mqttpassed

  ### Session Options
  #qos: 0
  #retain: false

  ### Topics
  # Substitutions:
  # - {hostname} = the computer's hostname
  # - {client_id} = the mqtt 'client id'
  # topics are composed as `{root topic}/{node id}/{property id}`
  #root topic: '{client_id}' # Root topic
  # last will topic becomes `{root topic}/{lwt topic}`
  #lwt topic: 'lwt' # Last Will topic

  ### Heartbeat
  # Interval in seconds to publish status messages
  #heartbeat interval: 300


######## GPIO Module ########
gpio:

  ### Pin Numbering Mode
  #mode: 'SOC'

  ### Polling Interval
  #polling interval: 60 # interval in seconds to poll all pins

  ### Interrupt Debounce
  #debounce: 50 # in milliseconds

  ### Single Pin Configuration
  pin-id:

    ### Pin Number
    pin: 4

    ### Pin Name
    # Subsitutions:
    # - {pin} = pin number
    # - {pin:02d} = pin number padded with 0's to always be 2 numbers
    # - {pin_id} = name of this section
    # name: '{pin_id}'

    ### Pin Mode
    pin mode: 'input' # can be 'input' or 'output'

    ### Pin Resistor
    #resistor: 'off' # can be 'pullup', 'pulldown', or 'off'

    ### Pin Type Specifc Options
    # Each pin type will have some options specific to it,
    # consult the documentation for details.

  ### Multiple Pin Configuration
  # alternatively a list of pins can be given.
  # actual id's will be '{batch-id}-{index + first index}'
  # all options from a single pin configuration apply, except those shown below
  #batch-id:

    ### Pin Name
    # a single name may be given, or a list of names the same length as the
    # list of pins can be provided. if a single name is given then actual names will
    # be '{name} {index + first index}'
    # Subsitutions:
    # - {pin} = pin number
    # - {pin:02d} = pin number padded with 0's to always be 2 numbers
    # - {pin_id} = name of this section
    # - {index} = the pin's index in the list of pins plus 'first index'
    #name: '{pin_id}'

    ### Pin Numbers
    #pin: []

    ### First Index
    # this number will be added to 'index' when substituting '{index}' in id's and names
    #first index: 0



######## I2C Module ########
#i2c:

  ### Polling Interval
  #polling interval: 60 # interval in seconds to poll all pins

  ### Device Configuration
  # !! this id is used as the node so it must not match any other node or module name !!
  #device-id:

    ### Device Friendly Name
    # Friendly name used in logging etc
    # Substitutions:
    # - {device_id} = name of device's section in config file
    # - {address:02x} = device address in hex (ex. '20')
    # - {device} = device identifier
    #name: '{device_id}'

    ### Device Type
    # Device identifier, see the wiki for a list of supported devices
    #device: ''

    ### I2C Bus ID
    # ID of the I2C bus to use on your SBC
    # ex. Raspberry Pi 3 this will probably be 1
    # Can also specify full path ex. '/dev/i2c-1'
    #bus:

    ### Device Address
    # 7-bit hex address of the device (ex. 0x20)
    # this can also be a base-10 integer
    #address: 0x

    ### Device Specific Options
    # Each device type will have some options specific to it
    # consult the wiki page for the device for details.



######## OneWire Module ########
#onewire:

  ### Bus Interface
  #bus: 'w1'

  ### Polling Interval
  #polling interval: 60 # interval in seconds to poll all devices

  ### Bus Scan
  # Scan bus for any unconfigured devices and add them
  #bus scan: false

  ### Device Configuration
  #device-id:

    ### Device Name
    # Substitutions:
    # - {device_id} = will be replaced with the device definition section name
    # - {device_type} = will be replaced with the device type (ex. `DS18B20`)
    # - {address} = will be replaced with the device's **FULL 8 BYTE ADDRESS**
    #name: '{device_id}'

    ### Device Address
    # The device's address, can be 7 or 8 bytes, can use w1 style `xx-xxxxxxxxxxxx`
    #address: '' # **MUST BE IN SINGLE QUOTES `''`**

    ### Device Specific Options
    # Each device type will have some options specific to it
    # consult the wiki page for the device for details.

  ### Multiple Device Configuration
  #batch-id:

    ### Device Name
    # a single name may be given, or a list of names the same length as the
    # list of pins can be provided. if a single name is given then actual names will
    # be '{name} {index + first index}'
    # Substitutions:
    # - {device_id} = will be replaced with the device definition section name
    # - {device_type} = will be replaced with the device type (ex. `DS18B20`)
    # - {address} = will be replaced with the device's **FULL 8 BYTE ADDRESS**
    # - {index} = will be replaced with the device's index in the list of addresses
    #name: '{device_id}'

    ### Device Address
    # List of device addresses, can be 7 or 8 bytes, can use w1 style `xx-xxxxxxxxxxxx`
    #address: ['', ''] # **MUST BE IN SINGLE QUOTES `''`**

    ### First Index
    # this number will be added to 'index' when substituting '{index}' in id's and names
    #first index: 0



######## LED Module ########
#led:

  ### Animation Directory
  # A directory or list of directories to look for additional animations in
  #anim dir: []

  ### Startup Animation
  # Animation to play when MQTTany loads
  #anim startup: 'test.array'

  ### Array Configuration
  #array-id:

    ### Array Name
    # Substitutions:
    # - {array_id} = will be replaced with the array definition section name
    #name: '{array_id}'

    ### Output
    # The interface to use for outputting data
    # Available options are 'rpi' or 'sacn'
    #output: ''

    ### Pixel Count
    # The number of pixels in your array (not necessarily LEDs, see next option)
    #count:

    ### LEDs per Pixel
    # Allows you to have more than one LED handled as a single pixel.
    # Ex. A bulb with 3 LEDs in it can be treated as a single pixel in the array.
    # Total number of LEDs in the array should be 'count' times 'leds per pixel'
    #leds per pixel: 1

    ### Initial Brightness
    # The initial brightness of the array, can be 0-255
    #brightness: 255

    ### Color Order
    # The byte order for the color data sent to the LEDs.
    # See the 'test.order' animation for how to determine this
    #color order: '{default}'

    ### Animation Frame Rate
    # Frame rate for animations, you may need to decrease this for
    # longer arrays or certain interface types.
    #anim fps: 60

    ### Interface Specific Options
    # Each interface may have some options specific to it
    # consult the wiki page for the interface for details.



######## XSET Module ########
#xset:

  ### Default Display
  # Display option to use when commands do not specify a display.
  # Omitting this option will omit the '-display' option from the call to xset
  #default display: None

  ### Startup Commands
  # A list of commands to run when the module loads.
  # See the documentation on sending commands for formatting.
  # If none are desired, use an empty list as shown below.
  #startup commands: []
***@***.***:~ $ 

thanks in advance 

bedst regards 

John

**Comment by CrazyIvan359** [comment]
Sorry I didn't see your reply here.

Are you running Armbian? If so, check what interfaces are enabled using `armbian-config` as it looks like OneWire is enabled. `system-1wire` is not part of MQTTany, that is the OneWire driver which means it's turned on somewhere, event if it's not in `boot.txt`.

**Reply by guitarddns** [reply]
Thanks for your reply 
I have dobbelt check the "raspi-config" and 
The one-wire interface is disabled

is it a reinstall of the raspberry that need here - or do I have a misconfig of the gpio part of the mqttany.yml file 
thanks in advance 
John

**Reply by guitarddns** [reply]
A new update 
I changed the pin from 4 to 5 and it came up like expected
It look like pin 4 is the default for one-wire - and even one-wire is not load 
__________________________________________________________________________
root@Infra-read:/home/john# lsmod
Module                  Size  Used by
sha256_generic         16384  0
cfg80211              761856  0
rfkill                 32768  2 cfg80211
8021q                  32768  0
garp                   16384  1 8021q
stp                    16384  1 garp
llc                    16384  2 garp,stp
raspberrypi_hwmon      16384  0
bcm2835_codec          40960  0
bcm2835_v4l2           40960  0
bcm2835_isp            32768  0
v4l2_mem2mem           36864  1 bcm2835_codec
bcm2835_mmal_vchiq     32768  3 bcm2835_isp,bcm2835_codec,bcm2835_v4l2
videobuf2_vmalloc      16384  1 bcm2835_v4l2
videobuf2_dma_contig    20480  2 bcm2835_isp,bcm2835_codec
videobuf2_memops       16384  2 videobuf2_dma_contig,videobuf2_vmalloc
snd_bcm2835            24576  1
videobuf2_v4l2         32768  4 bcm2835_isp,bcm2835_codec,bcm2835_v4l2,v4l2_mem2mem
videobuf2_common       61440  5 bcm2835_isp,bcm2835_codec,bcm2835_v4l2,v4l2_mem2mem,videobuf2_v4l2
snd_pcm               110592  1 snd_bcm2835
videodev              249856  6 bcm2835_isp,bcm2835_codec,videobuf2_common,bcm2835_v4l2,v4l2_mem2mem,videobuf2_v4l2
snd_timer              32768  1 snd_pcm
mc                     45056  6 bcm2835_isp,bcm2835_codec,videobuf2_common,videodev,v4l2_mem2mem,videobuf2_v4l2
snd                    77824  5 snd_timer,snd_bcm2835,snd_pcm
vc_sm_cma              32768  2 bcm2835_isp,bcm2835_mmal_vchiq
uio_pdrv_genirq        16384  0
uio                    20480  1 uio_pdrv_genirq
fixed                  16384  0
ip_tables              28672  0
x_tables               32768  1 ip_tables
ipv6                  495616  29
root@Infra-read:/home/john#

and no entry in /boot/config.txt or raspi-config 
it still continue to complain - until I change the mqttany.yml config file from pin 4 to pin 5


__________________________________________________________________________
â mqttany.service - MQTTany service
   Loaded: loaded (/etc/systemd/system/mqttany.service; enabled; vendor preset: enabled)
   Active: active (running) since Tue 2023-01-24 13:48:41 GMT; 3s ago
 Main PID: 1159 (python3)
    Tasks: 10 (limit: 1935)
   CGroup: /system.slice/mqttany.service
           ââ1159 /usr/bin/python3 mqttany.py
           ââ1169 /usr/bin/python3 mqttany.py
           ââ1170 /usr/bin/python3 mqttany.py

Jan 24 13:48:42 Infra-read python3[1159]: 2023-01-24 13:48:42,532 [WARN ] [core                    ] #  This is a development version and  #
Jan 24 13:48:42 Infra-read python3[1159]: 2023-01-24 13:48:42,533 [WARN ] [core                    ] #    may be incomplete or unstable    #
Jan 24 13:48:42 Infra-read python3[1159]: 2023-01-24 13:48:42,535 [WARN ] [core                    ] #                                     #
Jan 24 13:48:42 Infra-read python3[1159]: 2023-01-24 13:48:42,536 [WARN ] [core                    ] #######################################
Jan 24 13:48:42 Infra-read python3[1159]: 2023-01-24 13:48:42,538 [WARN ] [core                    ]
Jan 24 13:48:42 Infra-read python3[1159]: 2023-01-24 13:48:42,784 [INFO ] [core.gpio               ] Detected board: RASPBERRY_PI_2B
Jan 24 13:48:43 Infra-read python3[1159]: 2023-01-24 13:48:43,358 [INFO ] [core                    ] Module 'mqtt' started successfully
Jan 24 13:48:43 Infra-read python3[1159]: 2023-01-24 13:48:43,371 [INFO ] [core                    ] Module 'gpio' started successfully
Jan 24 13:48:43 Infra-read python3[1159]: 2023-01-24 13:48:43,386 [INFO ] [gpio.digital            ] Setting up 'pin-id' on GPIO05 as OUTPUT

**Reply by CrazyIvan359** [reply]
Puzzling indeed. If possible I would suggest using another pin for now.

Have you enabled OneWire on this Pi and then disabled it, or has it never been enabled since you installed Raspbian?

I don't have an actual Pi at hand right now to test this on, but I am going to open an issue and look into it when I have some time.
